[Unified Search] Supports complex filters with AND/OR relationships

nlatipov opened the pull request:
## Describe the feature:
Closes https://github.com/elastic/kibana/issues/144775

This PR allows users to create more than one filter at a time. It enhances the query builder by enabling users to create multiple filters simultaneously. It adds the capability to nest queries and use the logical OR operator in filter pills.

<img width="981" alt="image" src="https://user-images.githubusercontent.com/4016496/207942022-3256590d-00f6-45c8-b566-c184d5d18953.png">

## Tasks:

- [x] Add the ability to add/edit multiple filters in one form:
  - [x] Replace the current implementation of adding and editing a filter with a filtersBuilder - `Vis-Editor`;
  - [x] Add combined filter support to Data plugin (mapAndFlattenFilters) - `App-Services`;
  - [x] Add the ability to update data in the Data plugin when updating values ​​in the filters builder - `App-Services`;
  - [x] Add hide `Edit as Query DSL` in popover case the filter inside FiltersBuilder is combinedFilter  - `App-Services`;
- [x] Update filter badge to display nested filters:
  - [x] Replace the current badge filter implementation with a new one - `Vis-Editor`;
  - [x] Clean up `FilterLabel` component after replace `FIlterBadge` component - `Vis-Editor`;
- [x] When editing filters, those filters that belong to the same filter group should be edited - `Vis-Editor`;
- [x] Update jest and functional tests with new functionality - `Vis-Editor`;
- [x] Fix drag and drop behavior - `Vis-Editor`;

lukasolson committed:
Update combined filter to have relation type

nlatipov committed:
feat: updated combined filter supporting

nlatipov committed:
fix: type check, jest, comments

nlatipov committed:
fix: some test

nlatipov committed:
refact: conditionType -> booleanRelation

lukasolson committed:
Merge branch 'main' into combined-filter-relation

nlatipov commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into combined-filter-relation

lukasolson committed:
mapFilter support

lukasolson committed:
Fix DSL to wrap in bool

nlatipov committed:
fix: show error message

nlatipov committed:
fix: drag and drop in popover

nlatipov committed:
fix: display combinedFilter in filterBadge

nlatipov committed:
fix: editing combined filter

nlatipov committed:
fix: display filter item in filters builder

lukasolson committed:
Update buildCombinedFilter to accept additional filter args

nlatipov committed:
feat: hide toggle Edit as Query DSL for combinedFilter

nlatipov committed:
doc: added comment about popover drag and drop style

nlatipov committed:
fix: function tests in filter popover

nlatipov committed:
refact: rollback shouldShowToggle

nlatipov committed:
fix: creating combinedFilter

nlatipov committed:
fix: map combined test

nlatipov committed:
refact: little refact with filters builder

alexwizp committed:
try to fix CI

alexwizp committed:
remove extra translations

alexwizp committed:
test CI

alexwizp commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into combined-filter-relation

alexwizp committed:
fix a11 tests

nlatipov committed:
refact: clean up filterLabel

nlatipov committed:
fix: check type test CI

nlatipov commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into combined-filter-relation

nlatipov committed:
fix: type check test

nlatipov committed:
fix: linting test

alexwizp committed:
fix compresed

alexwizp committed:
revert wrong fix

nlatipov committed:
fix: autocomplite

alexwizp committed:
fix CI

nlatipov committed:
feat: added hide edit Query DSL

alexwizp committed:
fix showing errors

nlatipov committed:
feat: added dissabled toggle DSL in popover

nlatipov committed:
fix: changed message for dissabled DSL

alexwizp committed:
update placeholders

alexwizp committed:
remove createCustomLabel switch

alexwizp committed:
fix issue with updating operator

alexwizp commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into combined-filter-relation

alexwizp committed:
fix some CI steps

alexwizp committed:
fix i18n issue / fix typo

alexwizp commented on src/plugins/unified_search/public/filter_bar/filter_editor/lib/filter_editor_utils.ts:
```diff
@@ -63,7 +63,7 @@ export function isFilterValid(
   params?: any
 ) {
-  if (!indexPattern || !field || !operator) {
+  if (!indexPattern || !field || !operator || !params) {
```
@nlatipov please doube check that change.  I just see that for 'exist' filter you did 

<img width="607" alt="image" src="https://user-images.githubusercontent.com/20072247/199110296-b0c25d2a-877e-44ba-8576-eb9a3f395ad4.png">

alexwizp replied:
I've temporary removed that, if we still need that for some reason, let's discuss it offline.

nlatipov replied:
Thanks for bringing this matter to your attention.
I added an additional check 

```
  if (operator.type !== FILTERS.EXISTS && !params) {
    return false;
  }
```

Now if the operator is existential the check will be correct. And also the check will remain correct for other cases

alexwizp committed:
fix message

alexwizp committed:
revert wrong changes

andreadelrio commented:
Hello, I'm sending the first round of design feedback. I know the PR is not marked ready for review but I want to be proactive in providing design input.  (cc @stratoula)

<img width="862" alt="Frame 1" src="https://user-images.githubusercontent.com/4016496/199131137-79a88bff-a8e4-4674-ab29-5570a93fb972.png">

nlatipov committed:
fix: jest test for filter exists case

nlatipov committed:
fix: drag and drop in popover

stratoula committed:
Merge branch 'main' into combined-filter-relation

stratoula commented:
@andreadelrio thanx for your review. One question for you. On this PR we also change the filter badges. We are actually adding colors (blue for operations AND|OR and green for values)

I see that this change has also affected the popover that appears when I click a chart

<img width="420" alt="image" src="https://user-images.githubusercontent.com/17003240/199190920-a6d0e370-6bad-4e81-9c96-b40d76c92349.png">

You can see the values are green now. Are we ok with this change? Personally I am fine but I want your thoughts on that

nlatipov committed:
refact: isFilterValid

alexwizp committed:
fix labels

nlatipov committed:
feat: add icon for filter field

nlatipov committed:
refact: clean up filters builder filter item field input

nlatipov committed:
fix: filter badge show AND for OR combinedFilter

nlatipov committed:
fix: display prefix for combinedFilter

andreadelrio commented:
> @andreadelrio thanx for your review. One question for you. On this PR we also change the filter badges. We are actually adding colors (blue for operations AND|OR and green for values)
> 
> I see that this change has also affected the popover that appears when I click a chart
> 
> <img alt="image" width="420" src="https://user-images.githubusercontent.com/17003240/199190920-a6d0e370-6bad-4e81-9c96-b40d76c92349.png">
> 
> You can see the values are green now. Are we ok with this change? Personally I am fine but I want your thoughts on that

@stratoula I think we're good to keep the values green in the modal too.

stratoula committed:
Merge branch 'main' into combined-filter-relation

alexwizp committed:
fix Functional Tests

nlatipov committed:
feat: refact filterBadge and fix function test

alexwizp commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into combined-filter-relation

andreadelrio commented:
Stratoula and I discussed updating the AND/OR buttons to this:

<img width="833" alt="Frame 1" src="https://user-images.githubusercontent.com/4016496/199586117-a5bcc495-2fdd-4df6-b159-d5920cba518c.png">

You can get this by doing:

```
<EuiButtonEmpty
 iconType="plusInCircle"
 size="s"
 iconSize="s">
 OR
</EuiButtonEmpty>
```

Note: these buttons might change in the future as we're about to conduct usability testing but based on feedback we've already received we feel these are more clear than using the `plus` and the `returnKey` icons.

nlatipov committed:
style: updated OR and AND buttons

nlatipov committed:
style: filter edit popover

andreadelrio commented:
As discussed in the bi-weekly meeting today, let's also add a horizontal line for the operation AND (just like the one we have for OR). This will help with readability.

<img width="810" alt="image" src="https://user-images.githubusercontent.com/4016496/199850059-d051e971-1b5e-462a-93ad-230055bd961b.png">

lukasolson committed:
Fix issue with compareFilter with combined filters

lukasolson committed:
Merge branch 'combined-filter-relation' of github.com:elastic/kibana into combined-filter-relation

stratoula commented:
Hey guys some bugs I found
- Adding the value on a boolean field is not allowed
<img width="831" alt="image" src="https://user-images.githubusercontent.com/17003240/199923500-45be807b-edec-4165-bd4b-c40b826d1715.png">
- Editing works if I just change the value but it doesn't if I add another row of filter or if I drag and drop a row on another drop target.

alexwizp committed:
Merge remote-tracking branch 'upstream/main' into combined-filter-relation

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/generate codeowners'

nlatipov committed:
refact: style filters builder

alexwizp committed:
fix for Boolean fields

alexwizp committed:
preparation utils for moving into es-query

alexwizp commented:
### Issues 

- [ ] The component should work correctly when the `DataView` changes

https://user-images.githubusercontent.com/20072247/200010173-e4eaf371-c9ed-4de0-8000-d1f847e7136c.mov

- [ ] We should clean filters in case of switching `DataView` 

https://user-images.githubusercontent.com/20072247/200011039-4bd3a903-79a5-4a36-aae9-94cf63d3978a.mov


### Design Questions

- [ ]  @andreadelrio I think we need a placeholder for cases when `DataView` is not selected.

<img width="855" alt="image" src="https://user-images.githubusercontent.com/20072247/200011920-9d0701d0-3bd5-4ecc-8aae-2bdc80a985bb.png">
 
- [ ] @andreadelrio / @stratoula  I'm not a designer but for cases with DataView selector it looks strange

<img width="856" alt="image" src="https://user-images.githubusercontent.com/20072247/200012949-f9be10d9-c99b-4757-bbb0-0d53b046a929.png">

stratoula commented:
Yes I agree @alexwizp, @andreadelrio maybe we could also improve the dataview picker too? It appears on the dashboard when you have viz from different dataviews.

andreadelrio commented:
> ### Design Questions
> * [ ]   @andreadelrio I think we need a placeholder for cases when `DataView` is not selected.
> 
> <img alt="image" width="855" src="https://user-images.githubusercontent.com/20072247/200011920-9d0701d0-3bd5-4ecc-8aae-2bdc80a985bb.png">
> 

In this scenario, is it possible to show the inputs (field + operator + value(s)) disabled and on hover show a tooltip saying "You need to select a data view first"? cc @alexwizp

nlatipov committed:
style: chened width for fitler item and fix style for list

nlatipov committed:
fix: show empty fields in filter popover if dataView changed

nlatipov committed:
fix: updated combinedFilter

nlatipov commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into combined-filter-relation

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/generate codeowners'

andreadelrio commented:
It seems like we've inverted the treatment we give to a click on "OR" button with the one we give to a click on the "AND" button. Before, clicking on the "OR" button would create a nested box with white background. Can we revert this? (cc @stratoula )

![and_or_inverted](https://user-images.githubusercontent.com/4016496/200406469-2435d21c-d30e-4a15-b617-a8895c9ad55f.gif)

stratoula commented:
Also team we would like to add a preview functionality on the popover. Check how this should look

https://user-images.githubusercontent.com/17003240/200497567-ba0fced7-e9f7-44ae-b864-b8eb4d23280a.mov

stratoula commented:
@andreadelrio we are a bit confused about the custom label. How this is going to look like? As it is right now (in main) with the switch? Are we changing it to display the input without the switch? Are we going to take this decision after the usability testing?

andreadelrio commented:
> @andreadelrio we are a bit confused about the custom label. How this is going to look like? As it is right now (in main) with the switch? Are we changing it to display the input without the switch? Are we going to take this decision after the usability testing?

@stratoula I think it should like in `main` (with the EuiSwitch) but we can wait after the first round of usability testing. Irrespective of how we display it, please place it below the `Preview`.

andreadelrio commented:
Hi team, some more design considerations:

- For the Preview we can use the `inspect` iconType (which we also use in the "Create field" preview window in Discover).
- We should bump up the maximum height of the query building section `filtersBuilderMaxHeight` to at least 300px.
- We can optimize vertical space by adjusting the padding, particularly in the nested filter box. Here's a couple of suggestions

<img width="657" alt="Frame 2" src="https://user-images.githubusercontent.com/4016496/200697975-55462709-e3eb-4700-b7ed-056e4b709847.png">

nlatipov committed:
style: update design

nlatipov committed:
feat: added filterBadge into filter editor

nlatipov committed:
fix: type check

alexwizp committed:
add FilterBadgeErrorBoundary component

alexwizp commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into combined-filter-relation

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/generate codeowners'

alexwizp commented:
@andreadelrio I'm just thinking how the "preview" panel should looks "by default" / "in case of one inner filter is not valid". 

For now we implemented that in that way: 


<img width="878" alt="image" src="https://user-images.githubusercontent.com/20072247/200817335-6a83851a-49dc-4ac5-b9da-5105eea46a0a.png">

<img width="868" alt="image" src="https://user-images.githubusercontent.com/20072247/200817456-f621c5c6-256b-4c17-93bd-20ec3b2cb1e4.png">




- [ ] @Nodir in case of `AND` operation for some reason the delimiter is not shown. Please fix  
        

> <img width="527" alt="image" src="https://user-images.githubusercontent.com/20072247/200818448-5139b073-d11c-46f3-b07f-25542973f187.png">

alexwizp committed:
some cleanup

stratoula commented:
Can we give more space here among the lines of the preview?

<img width="819" alt="image" src="https://user-images.githubusercontent.com/17003240/200838998-796377d5-c494-478c-bf3a-3a8399c425d0.png">

andreadelrio commented:
> @andreadelrio I'm just thinking how the "preview" panel should looks `by default` / `in case of one inner filter is not valid`.
> 
> For now we implemented that in that way:
> 
> <img alt="image" width="878" src="https://user-images.githubusercontent.com/20072247/200817335-6a83851a-49dc-4ac5-b9da-5105eea46a0a.png">

When there's only one row and they haven't made any selections we shouldn't show the `Preview` panel at all. Otherwise, it's like we're telling the users that they've done something wrong when they just opened the form. The disabled `Add Filter` button should be enough to indicate to them that they need to make selections. Other than that, I think the subdued badge for invalid/incomplete works but let's see what comes out from usability testing and if we should refine in any way.

andreadelrio commented:
> Can we give more space here among the lines of the preview?
> 
> <img alt="image" width="819" src="https://user-images.githubusercontent.com/17003240/200838998-796377d5-c494-478c-bf3a-3a8399c425d0.png">

++ and I think we're fine to reduce the text 12px.

alexwizp commented:
@andreadelrio could you please give us some recommendations for small screens

<img width="628" alt="image" src="https://user-images.githubusercontent.com/20072247/200903979-5a391db4-006a-41a1-96e1-0a6b4678885e.png">

alexwizp committed:
do some cleanup

nlatipov committed:
style: updated for filter badge in filter editor by design suggestion

alexwizp committed:
update "shouldShowPreview" condition

alexwizp committed:
fix brackets for negate combined filter /  fix flickering when displaying panel preview

alexwizp committed:
fix issue on changing data view

alexwizp committed:
Merge remote-tracking branch 'upstream/main' into combined-filter-relation

nlatipov committed:
style: fix regres with warning message in filterBadge

nlatipov committed:
fix: show empty fitler with dataView didn't select

alexwizp committed:
do some cleanup

alexwizp committed:
remove strange code

alexwizp committed:
Merge remote-tracking branch 'upstream/main' into combined-filter-relation

alexwizp committed:
remove uiSettings / InjectedIntl for FIlterEditor

nlatipov committed:
style: fix popover view with small screen

alexwizp committed:
update jest

alexwizp committed:
fix pinned/negate/disable for filters

nlatipov committed:
style: filter builder group for small screen

alexwizp committed:
fix range filter / fix edit DSL

nlatipov committed:
fix: added space between label and value in filter badge

nlatipov committed:
fix: updated snapshot for filter content

nlatipov commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into combined-filter-relation

nlatipov commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into combined-filter-relation

alexwizp committed:
fix fileter validation

alexwizp committed:
fix filter_editor

nlatipov committed:
feat: added disabled component if dataView didn't select

nlatipov committed:
fix: setting negative value

alexwizp committed:
remove updateOperator

alexwizp commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into combined-filter-relation

alexwizp commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into combined-filter-relation

nlatipov committed:
refact: remove hardcode filter editor with size element

alexwizp commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into combined-filter-relation

Kuznietsov committed:
D'n'd order fix.

Kuznietsov committed:
handleCombinedFilter refactoring.

Kuznietsov committed:
Renamed handle_combined_filter.ts

Kuznietsov committed:
Renamed handleNestedFilter.

Kuznietsov committed:
Renamed handle_nested_filter.ts

Kuznietsov committed:
excludedAttributes cleanup refactored.

Kuznietsov committed:
FilterContent export fixed.

Kuznietsov committed:
FilterBadge refactored.

Kuznietsov committed:
FilterExpressionBadge refactored.

Kuznietsov committed:
FilterBadgeGroup refactored.

Kuznietsov committed:
Refactored styles and strings.

Kuznietsov committed:
Removed unused export of DEFAULT_MAX_HEIGHT.

Kuznietsov committed:
Fixed getFieldFromFilter type.

Kuznietsov committed:
Renamed doRecursive.

Kuznietsov committed:
Renamed flags for uniformity.

Kuznietsov committed:
filter_item refactored.

Kuznietsov committed:
FilterGroup refactored.

Kuznietsov committed:
Moved strings.ts to i18n.ts

Kuznietsov committed:
Added type safety for params_editor methods and refactored.

andreadelrio commented:
Design feedback - Nov 22, 2022

- [ ] Remove the drag handle and the ability to drag (and drop) when there's only 1 row as they're unnecessary.

<img width="845" alt="image" src="https://user-images.githubusercontent.com/4016496/203443768-7f234ad3-5902-4451-a387-1597246107b8.png">
Proposed design for 1 row:
<img width="844" alt="image" src="https://user-images.githubusercontent.com/4016496/203444770-724400e7-9b5d-4226-a92f-f656004b5720.png">

- [ ] Don't close the form when clicking outside of it, that might lead to losing work in progress
![Screen Recording 2022-11-22 at 5 51 04 PM](https://user-images.githubusercontent.com/4016496/203455979-9407c8d4-a9ed-4d77-8474-d51dc7c7d484.gif)

Kuznietsov committed:
Fixed bug.

Kuznietsov committed:
Moved styles.

Kuznietsov committed:
FiltersBuilder refactoring.

alexwizp commented:
@andreadelrio @stratoula  
Have a question about  

> Remove the drag handle and the ability to drag (and drop) when there's only 1 row as they're unnecessary.

I think it would be better not to `delete`, but to `disable`, since in other places we use this particular pattern e.g. in Lens

https://user-images.githubusercontent.com/20072247/203530995-3d27ed08-0b38-4f08-861d-fe19ff22d00b.mov

Kuznietsov committed:
updateFilter refactored.

Kuznietsov committed:
Moved AddFilterPopover styles.

Kuznietsov committed:
Fixed i18n and styles.

stratoula commented:
@alexwizp yes I think I prefer this approach better. We are aligned with lens behavior and the form doesn't change when I have multiple rows. 👍

Kuznietsov committed:
Moved translations from query_string_input to i18n.ts file.

Kuznietsov committed:
Reduced the size of unified_search bundle by moving strings to corresponding files.

andreadelrio commented:
> @andreadelrio @stratoula Have a question about
> 
> > Remove the drag handle and the ability to drag (and drop) when there's only 1 row as they're unnecessary.
> 
> I think it would be better not to `delete`, but to `disable`, since in other places we use this particular pattern e.g. in Lens
> 

@alexwizp Agreed! this is cleaner. Let's include a tooltip too.

Kuznietsov committed:
Fixed the problem with normalization.

Kuznietsov committed:
Fixed combined filter bug.

Kuznietsov committed:
styles bug fixed.

Kuznietsov committed:
fixed custom label.

Kuznietsov committed:
Fixed FilterItem actions on small screen.

Kuznietsov committed:
Added small buttons for 's'.

Kuznietsov committed:
Disabled dragging if one element is added only.

Kuznietsov committed:
Fixed `is on of` input while pasting non-existence value.

Kunzetsov commented:
> ### **Design feedback - Nov 22, 2022**
> ...
> * [ ]  Don't close the form when clicking outside of it, that might lead to losing work in progress
>   ![Screen Recording 2022-11-22 at 5 51 04 PM](https://user-images.githubusercontent.com/4016496/203455979-9407c8d4-a9ed-4d77-8474-d51dc7c7d484.gif)
>         

@andreadelrio, Thanks for this suggestion. Unfortunately, it can lead to the unexpected behavior of the `eui` library.

Normal behavior: 

https://user-images.githubusercontent.com/22456368/203819921-685c87a9-39c1-4b37-a272-8d9e5b8ec168.mov

Unexpected behavior (not responding anymore + strange behavior even if it would work properly):

https://user-images.githubusercontent.com/22456368/203820174-df413a69-e4ef-4f6b-95b0-10299078c010.mov


WDYT about it? Possibly, it is better to add a confirmation popup. What to hear your opinion on that. Thanks 😃

Kuznietsov committed:
createFilter utility for functional tests added.

Kuznietsov committed:
Replaced old implementation of createFilter with new one.

Kuznietsov committed:
Fixed one more test.

Kuznietsov committed:
Fixed dashboard tests.

Kunzetsov commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into combined-filter-relation

Kuznietsov committed:
Fixed incomming test.

Kuznietsov committed:
Merge branch 'main' into combined-filter-relation

Kuznietsov committed:
One more test fixed.

Kunzetsov commented:
@stratoula, right now we have this title of the filter. It has a bug, related to the `OR` and `AND` without spaces around. The title is taken as an inner text of the Component, which is printing the filter in some way. WDYT about the total removal of the filter's content from the title?

List of the proposed titles:

- `Select for more filter actions`; 
- `Disabled filter. Select for more filter actions`;
- `Exclude results. Select for more filter actions.`;
- `Disabled filter. Exclude results. Select for more filter actions.`;

<img width="450" alt="Screenshot 2022-11-28 at 12 17 15" src="https://user-images.githubusercontent.com/22456368/204252751-9a2f4110-7e72-4129-a8f5-f1fdf0c078d6.png">

Kuznietsov committed:
Fixed issue with at title and added data-test-subj for complex filters.

stratoula commented:
@Kunzetsov if we can't fix it and show the complex filter correctly then I am fine to proceed like this.

Kuznietsov committed:
Add or filter test added.

Kuznietsov committed:
Functional tests for AND filter and nested filters added.

Kuznietsov committed:
Fixed filter-value-${value} for values with spaces.

Kuznietsov committed:
Fixed jest test.

Kuznietsov committed:
Fixed cypress tests.

Kuznietsov committed:
Fixed selected_filters test.

Kuznietsov committed:
Possibly, fixed cypress tests.

elasticmachine commented:
Pinging @elastic/kibana-visualizations @elastic/kibana-visualizations-external (Team:Visualizations)

stratoula committed:
Merge branch 'main' into combined-filter-relation

Kuznietsov committed:
Merge branch 'main' into combined-filter-relation

ThomThomson commented:
Closing in favor of https://github.com/elastic/kibana/pull/146503#pullrequestreview-1197859329

Kuznietsov committed:
Merge branch 'main' into combined-filter-relation

banderror commented:
Hey @marshallmain and @elastic/security-detections-response-alerts, I added you as reviewers since it touches the rule forms on the Creation and Editing pages.

Kuznietsov committed:
Merge branch 'main' into combined-filter-relation

Kuznietsov committed:
euiBadge color changed to lightestShade.

Kuznietsov committed:
Fixed cursor on disabled drag n drop element.

Kuznietsov committed:
Fixed types.

nlatipov committed:
fix: drag and drop in query bar menu

Kuznietsov committed:
Merge branch 'main' into combined-filter-relation

Kuznietsov committed:
Fixed addFilter args.

Kuznietsov committed:
Added combobox with truncated option label.

Kuznietsov committed:
Small refactoring.

Kuznietsov committed:
Merge branch 'main' into combined-filter-relation

Kuznietsov committed:
Merge branch 'main' into combined-filter-relation

thomasneirynck commented:
fwiw - hitting this error 

![image](https://user-images.githubusercontent.com/1833023/205739410-05c0d6bd-856a-4480-a88e-187725990bb5.png)

Kuznietsov committed:
Fixed problem with Discovery.

stratoula reviewed (commented):
This looks fantastic! Some comments from my side:
- For some reason I am not allowed to add this filter and you can see on the preview tha tthe filter is invalid. It shouldn't though
<img width="848" alt="image" src="https://user-images.githubusercontent.com/17003240/205887375-03a26725-885a-42a7-873e-25e9a71fe407.png">
-Same here 
<img width="822" alt="image" src="https://user-images.githubusercontent.com/17003240/205887919-d81b12b6-33f7-463b-9472-0377178c1f78.png">

stratoula commented:
clicking on a chart to add a filter looks wrong

<img width="614" alt="image" src="https://user-images.githubusercontent.com/17003240/205889076-d4f4be55-eb8f-4ed8-b771-7f2e855171a7.png">

stratoula commented:
Creating a go to discover drilldown leads to Discover with the save icon on the filter badge while it shouldn't. I haven't saved this filter.

<img width="754" alt="image" src="https://user-images.githubusercontent.com/17003240/205889696-327c2ccf-8fd3-4a13-ad71-0b067fe7f3cd.png">

Kuznietsov committed:
Fixed regressions.

Kunzetsov commented on src/plugins/unified_search/public/filter_bar/filter_item/filter_item.tsx:
```diff
@@ -290,7 +310,7 @@ export function FilterItem(props: FilterItemProps) {
         {
           id: 'unifiedSearch.filter.filterBar.labelWarningInfo',
-          defaultMessage: 'Field {fieldName} does not exist in current view',
+          defaultMessage: 'Fields {fieldName} does not exist in current view',
```
Have no idea 😃 
Turned back the code.

Kuznietsov committed:
Fixed label.

Kuznietsov committed:
Fixed label of Lens filter on bar click.

xcrzx reviewed (commented):
On the detection rule view page, filters are not correctly displayed:

<img width="1564" alt="Screenshot 2022-12-06 at 14 18 42" src="https://user-images.githubusercontent.com/1938181/205923377-e9e98f4e-9a43-44c0-8b19-30462ec111f6.png">

xcrzx commented:
Also, I think there is a usability issue. Long field names are truncated, so it becomes impossible to read them.

<img width="1029" alt="Screenshot 2022-12-06 at 14 09 18" src="https://user-images.githubusercontent.com/1938181/205923620-c002a462-f792-4176-ac21-b3aaba555f1b.png">

Would it be possible to make the dropdown wider?

stratoula commented:
@xcrzx we know about the usability issue and we are exploring alternatives with @andreadelrio!

Kuznietsov committed:
Removed Saved icon on if alias to the filter is defined.

Kuznietsov committed:
Fixed tests.

Kunzetsov commented:
> On the detection rule view page, filters are not correctly displayed:
> 
> <img alt="Screenshot 2022-12-06 at 14 18 42" width="1564" src="https://user-images.githubusercontent.com/1938181/205923377-e9e98f4e-9a43-44c0-8b19-30462ec111f6.png">

@xcrzx, I'm sorry, could you, please, provide the steps to reproduce the behavior? Unfortunately, I couldn't do that.

Kuznietsov committed:
Fixed regression.

Kuznietsov committed:
Fixed Label at security_solution rules.

xcrzx commented:
> @xcrzx, WDYT, is it good enough?

Looks awesome, thank you!

xcrzx reviewed (approved):
Looked through the code changes and tested locally, LGTM!

Kuznietsov committed:
boxesHorizontal for mobile screen.

Kuznietsov committed:
Merge branch 'main' into combined-filter-relation

Kuznietsov committed:
Added max nesting of filter.

Kuznietsov committed:
Fixed paddings in mobile screen.

andreadelrio committed:
[Unified search] Complex form design cleanup (#147226)

## Summary

First pass at design code. Trying to get some more space in the form.

Updated form:

<img width="1026" alt="image"
src="https://user-images.githubusercontent.com/4016496/206339021-fcfeff88-087a-41fc-bd4a-7788f2054b02.png">

Kuznietsov committed:
Merge branch 'main' into combined-filter-relation

Kuznietsov committed:
Merge branch 'main' into combined-filter-relation

Kuznietsov committed:
Merge branch 'main' into combined-filter-relation

Kuznietsov committed:
Merge branch 'main' into combined-filter-relation

Kunzetsov commented:
@elastic/security-threat-hunting, could you, please, look at the cypress tests? There is a problem with `getAvailablePrebuiltRulesCount` method at `x-pack/plugins/security_solution/cypress/tasks/api_calls/prebuilt_rules.ts`. `Prebuild rules count is always zero and it is reproducible on the main. Could you, please, help with this issue? Thanks.

Kunzetsov commented:
@elastic/security-threat-hunting, could you, please, look at the cypress tests? There is a problem with `getAvailablePrebuiltRulesCount` method at `x-pack/plugins/security_solution/cypress/tasks/api_calls/prebuilt_rules.ts`. `Prebuild rules count is always zero and it is reproducible on the main. Could you, please, help with this issue? Thanks.

Kuznietsov committed:
Merge branch 'main' into combined-filter-relation

Kuznietsov committed:
Fixed translations.

Kunzetsov commented:
@elastic/kibana-app-services, could you, please, review this PR? Thanks 😃

Kunzetsov commented:
@elastic/kibana-gis, could you, please, review this PR? Thanks.

Kunzetsov commented:
@elastic/ml-ui, could you, please, review this PR? Thanks.

Kunzetsov commented:
@elastic/kibana-cloud-security-posture, could you, please, review this PR? Thanks.

jughosta reviewed (approved):
Great job! Awesome to have a way to build filters in UI!

Data Discovery changes LGTM 👍 

One case looks like a bug: reordering OR items via Drag and Drop converts them into AND items:
![Dec-12-2022 10-00-52](https://user-images.githubusercontent.com/1415710/207021757-e67771b6-aed4-4073-ab8d-d165101b09b1.gif)

banderror commented:
@nlatipov @Kunzetsov Hey folks, I stumbled upon this PR and I have two questions regarding the data model of the filters.

We use filters that are created by the UI components changed in this PR for detection rules in Security Solution on several pages. We store filters in Alerting Framework's rules which are saved objects. Two questions here:

1. Was the data model of the filters affected somehow by this PR? Or this is purely a UI change?
2. How could we make sure that whenever and if someone changes the filters model, that person is notified/ knows that they also have to update all the saved object types where the filters can be stored (including detection rules) with corresponding SO migrations? By any chance, is it already solved for other saved objects e.g. saved queries?

cc @xcrzx

ThomThomson reviewed (approved):
Dashboard functional test changes LGTM!

Also tested locally with some quick dashboard smoke tests and everything works as expected! Ran into a strange [reporting bug](https://github.com/elastic/kibana/issues/147397), but was also able to repro it on main, so that isn't relevant here.

Overall I tested:
- And relations / Or relations
- many filters
- pinned filters get appended to filters loaded from saved object.
- Saved query loading replaces saved object filters
- Dashboard unsaved changes appears when expected

One small nit:
I know you've been thinking through making the field input wider in order to avoid cutting off longer field names, but even some of the longer operator names are getting cut off at the moment with my setup.

<img width="950" alt="Screen Shot 2022-12-12 at 2 54 34 PM" src="https://user-images.githubusercontent.com/14276393/207164559-94df43d7-cf93-4831-8240-1f88825a5911.png">

Kunzetsov commented:
> pages. We store filters in Alerting Framework's rules which are saved objects. Two questions here:

This is purely the UI change. The previous UI component, which was adding filters, could support only filters with one nesting level. And to display them it was enough to use FilterLabel component. But now, filters can be nested and should be displayed by the component, which can render a tree structure 😃 

So, the previous structure (plain one) and the nested structure are still supported and no migrations are required.

stratoula commented:
> @nlatipov @Kunzetsov Hey folks, I stumbled upon this PR and I have two questions regarding the data model of the filters.
> 
> We use filters that are created by the UI components changed in this PR for detection rules in Security Solution on several pages. We store filters in Alerting Framework's rules which are saved objects. Two questions here:
> 
> 1. Was the data model of the filters affected somehow by this PR? Or this is purely a UI change?

They are not affected so this is why we haven't provided a migration. The simple filters are compatible with this version. This is an enhancement. 

> 2. How could we make sure that whenever and if someone changes the filters model, that person is notified/ knows that they also have to update all the saved object types where the filters can be stored (including detection rules) with corresponding SO migrations? By any chance, is it already solved for other saved objects e.g. saved queries?
> 
> cc @xcrzx

What we are doing in Lens that we also store filters in our SOs is that we run the filters migrations too. This is the PR of how we did this in Lens https://github.com/elastic/kibana/pull/120305. You should also run these migrations on your SO to ensure that you keep the correct schema. This is not necessary for this PR but have it in mind for the future.

andreadelrio committed:
[Complex filters] Prevent operator truncation  (#147409)

## Summary

Prevent the truncation of the selected operator by overriding EUI's
padding (temporary fix until permanent fix from EUI is released).

stratoula commented:
@Kunzetsov I have a dashboad with 2 visualizations that reference 2 different dataviews.

I select one dataview to create a filter and I see this error in the console

<img width="987" alt="image" src="https://user-images.githubusercontent.com/17003240/207258360-cd6f24b9-978d-4898-b64f-8fc00ad18085.png">
<img width="482" alt="image" src="https://user-images.githubusercontent.com/17003240/207258381-6551322a-e1c6-4436-a7bb-6bbdd8762636.png">

Can you check it?

Kuznietsov committed:
Fixed default max height for filters builder.

Kuznietsov committed:
Fixed cursor and fixed paddings on preview.

ppisljar commented:
```
{
		"meta": {
			"alias": "mmm",
			"disabled": false,
			"negate": false,
			"params": [
				{
				"$state": {
					"store": "appState"
				},
				"meta": {
					"alias": null,
					"disabled": false,
					"index": "ff959d40-b880-11e8-a6d9-e546fe2bba5f",
					"negate": false,
					"params": [{
						"$state": {
							"store": "appState"
						},
						"meta": {
							"alias": null,
							"disabled": false,
							"field": "category.keyword",
							"index": "ff959d40-b880-11e8-a6d9-e546fe2bba5f",
							"key": "category.keyword",
							"negate": false,
							"params": {
								"query": "Men's Shoes"
							},
							"type": "phrase"
						},
						"query": {
							"match_phrase": {
								"category.keyword": "Men's Shoes"
							}
						}
					}, {
						"$state": {
							"store": "appState"
						},
						"meta": {
							"alias": null,
							"disabled": false,
							"field": "category.keyword",
							"index": "ff959d40-b880-11e8-a6d9-e546fe2bba5f",
							"key": "category.keyword",
							"negate": false,
							"params": {
								"query": "Men's Accessories"
							},
							"type": "phrase"
						},
						"query": {
							"match_phrase": {
								"category.keyword": "Men's Accessories"
							}
						}
					}],
					"relation": "OR",
					"type": "combined"
				}
			},
```

seems like the $state property is not needed on nested filters and should only be there on the top level, same goes for disabled, negate, index pattern and alias.

stratoula commented:
@ppisljar this is a good point. I am just thinking that maybe in the future we might want to give the users the ability to disable specific parts of the nested filter or maybe negate too. About the index pattern, yes I agree we don't plan nested filters that reference different index pattern.

Kuznietsov committed:
Fixed problem with mutation of the filter from prop.

Kuznietsov committed:
Merge branch 'main' into combined-filter-relation

Kuznietsov committed:
Fixed comparison of filters.

Kuznietsov committed:
Merge branch 'combined-filter-relation' of github.com:elastic/kibana into combined-filter-relation

Kuznietsov committed:
Turned code back.

ppisljar commented:
> I am just thinking that maybe in the future we might want to give the users the ability to disable specific parts of the nested filter or maybe negate too
Unless we have that in the current plan I would not design for it as we can always add it later.

stratoula commented:
@ppisljar fair enough! @Kunzetsov let's do what Peter proposes

stratoula committed:
Merge branch 'main' into combined-filter-relation

Kuznietsov committed:
Merge branch 'main' into combined-filter-relation

stratoula commented:
When I add a Lens embeddable with complex filters I see this error on the console
<img width="1885" alt="image" src="https://user-images.githubusercontent.com/17003240/207531694-b03a283b-6e3b-4006-8cff-7cc39c5e74c5.png">

To reproduce it create a Lens chart with a complex filter i.e. 

<img width="1273" alt="image" src="https://user-images.githubusercontent.com/17003240/207531821-10f489ea-6091-4ba1-91b1-8eb6db9834ff.png">

Kuznietsov committed:
Fixed problem with assigning props to readonly filter.

Kuznietsov committed:
Quickfix of the failing filters.

Kuznietsov committed:
Removed index, negate, disabled, alias and $state from combined filter.

Kunzetsov commented on src/plugins/unified_search/public/filter_bar/filter_editor/lib/filter_editor_utils.ts:
```diff
@@ -20,7 +20,7 @@ export function getFieldFromFilter(filter: Filter, indexPattern?: DataView) {
 
 export function getOperatorFromFilter(filter: Filter) {
   return FILTER_OPERATORS.find((operator) => {
-    return filter.meta.type === operator.type && filter.meta.negate === operator.negate;
```
@ppisljar, can you tell me, please, why this part of the code has been added and if is it important? It is not working with the changes you requested.

ppisljar replied:
it tries to determine the correct operation from the meta properties. I think we should leave the negate part in there then as it seems it can be set in different ways than user negating the filter ? (cc @lukasolson )

lukasolson replied:
When creating a filter "is not" (or is not within range, etc.) we set the "negate" property, so this is required.

nreese reviewed (approved):
kibana-gis changes LGTM
code review - functional test changes only

ppisljar commented:
another issue, there is no query on the complex filter. I think there are a few places where we assume the query is there.

Kuznietsov committed:
Turned back negate.

Kuznietsov committed:
fixed one more test.

Kuznietsov committed:
Turned back index.

stratoula committed:
Merge branch 'main' into combined-filter-relation

Kuznietsov committed:
Added filters from searchSource in visualize_embeddable.

ppisljar commented on packages/kbn-es-query/src/filters/helpers/compare_filters.ts:
```diff
@@ -45,12 +45,15 @@ const mapFilter = (
-  if (comparators.negate) cleaned.negate = filter.meta && Boolean(filter.meta.negate);
-  if (comparators.disabled) cleaned.disabled = filter.meta && Boolean(filter.meta.disabled);
-  if (comparators.alias) cleaned.alias = filter.meta?.alias;
+  // Some filters are coming from GET params,
```
is this related to other changes in this commit ?

Kunzetsov replied:
Yes. These filters are without undefined values.
<img width="572" alt="Screenshot 2022-12-15 at 16 29 49" src="https://user-images.githubusercontent.com/22456368/207886680-48c66c31-8437-4143-95a6-2e506131bd97.png">

Kuznietsov committed:
Merge branch 'main' into combined-filter-relation

Kuznietsov committed:
Fixed problems with combined filters in TSVB/Agg based.

Kuznietsov committed:
Merge branch 'combined-filter-relation' of github.com:elastic/kibana into combined-filter-relation

lukasolson commented on src/plugins/data/common/search/expressions/filters_to_ast.ts:
```diff
@@ -6,21 +6,21 @@
-      ]).toAst();
-    });
+  return (Array.isArray(filters) ? filters : [filters]).map((filter) => {
+    const filterWithQuery = filter.meta.type === 'combined' ? fromCombinedFilter(filter) : filter;
```
Can you use `FILTERS.COMBINED` here instead of `'combined'?

Kunzetsov replied:
Done.

Kuznietsov committed:
Small refactoring.

Kunzetsov commented:
@elastic/ml-ui, could you, please, review this PR? Thanks.

stratoula reviewed (approved):
This looks great. Congrats for all this effort, it was hard I know! I tested:
- Visualize with complex filters and add this ti the dashboard
- TSVB  with complex filters and add this ti the dashboard
- Maps  with complex filters and add this ti the dashboard
- Vega  with complex filters and add this ti the dashboard
- Migration of dashboard with filters from 8.6
- Migration of saved query from 8.6
- Enable and disable the courier:ignore... setting
Everything seems to work fine.

Before merging can we only fix this https://github.com/elastic/kibana/pull/143928#issuecomment-1350535965?

Also @lukasolson and @ppisljar are you also ok?

stratoula commented:
Great find @walterra

In visualize works fine. Here is an unlinked saved search example, so I guess something goes wrong with the dataview in ml cc @Kunzetsov 

<img width="956" alt="image" src="https://user-images.githubusercontent.com/17003240/208088027-496cf0ef-26ac-4ec0-b7f6-edc869c72025.png">

alexwizp commented:
@walterra I think that the root cause of this problem lies in how the `ML` plugin works with `SavedSearch` saved object.   We debugged this code and found that instead of `indexPattern` obj, `indexPatternRef` gets into the filter. This means that the references were not injected.

Let's move on to the ML code and `x-pack/plugins/ml/public/application/util/index_utils.ts` file. I'm interested in the `getDataViewAndSavedSearch` and `loadSavedSearchById` methods. I see that we are returning and working with raw value which is not correct

My recommendation is to use the `getSavedSearch` method from `@kbn/saved-search-plugin/public`. Using this method will ensure that you have created and are using the correct `SavedSearch` type object. And since this is not related to this PR, let's do it as part of a new issue

plus I see that with one filter, this also does not work quite right 😢

Kuznietsov committed:
Fixed problems with modification.

Kuznietsov committed:
One more fix.

Kuznietsov committed:
Fixed tests.

qn895 commented:
> My recommendation is to use the `getSavedSearch` method from `@kbn/saved-search-plugin/public`. Using this method will ensure that you have created and are using the correct `SavedSearch` type object. And since this is not related to this PR, let's do it as part of a new issue

Thanks @alexwizp for the hint! I've created a new issue for ML [here](https://github.com/elastic/kibana/issues/147705) with the issues surrounding saved searches. Tested other functionalities of the filters (global pins) and ML changes LGTM.

andreadelrio reviewed (approved):
Design changes LGTM, great job! Looking forward to iterating as needed. Pending: add Technical Preview badge.

Kuznietsov committed:
Merge branch 'main' into combined-filter-relation

Kuznietsov committed:
Fixed test.

timductive commented:
Just adding a few comments here. Overall the experience is great, awesome work on this big PR.
I noticed a few minor nits, nothing blocking but perhaps some things to think about.

1. There's no autocomplete on the Value field in the screenshot below. Obviously, I can guess that bytes should be numeric but if I try to type alpha characters nothing happens. It would be nice to either, have the kql dropdown with some proposed values or at least an error message when I attempt to input an invalid value.
![image](https://user-images.githubusercontent.com/1280964/208186541-9c4d4ce2-d597-47da-aa46-158bbbb7831e.png)

2. I wonder if the "Edit filter" option should be at the top of the list here. I am much more frequently editing than pinning and I have to spend an extra second reading the options here.
![image](https://user-images.githubusercontent.com/1280964/208187146-15555b11-b13b-4dbb-8140-c0e75a36c1b2.png)

3. Sometimes the Preview shows the infinity symbol and sometimes its text
![image](https://user-images.githubusercontent.com/1280964/208187421-f1239726-1095-4ffc-b46e-85e2c558242e.png)
![image](https://user-images.githubusercontent.com/1280964/208187459-bdf93e16-1f87-4a0b-92f0-baba7144eac4.png)

4. Also, can we detect if the value is datetime and prompt the value with a datepicker?
![image](https://user-images.githubusercontent.com/1280964/208187660-85addc65-f3cd-454b-8d80-24764b7806ef.png)

5. There's some funkiness with pinning and switching between Discover and Lens. I can maybe recreate over zoom but if i just pin and unpin a bunch and switch between the apps, it seems like the state is getting confused..
Below in Lens i only have 1 pinned filter
![image](https://user-images.githubusercontent.com/1280964/208189043-35168113-8d84-401e-9693-ff48524d0440.png)
Then I switch to Discover and I have 2 pinned filters
![image](https://user-images.githubusercontent.com/1280964/208189145-61a23f9d-4aa4-4c39-81f0-45063ccb435a.png)

Kunzetsov commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into combined-filter-relation

Kunzetsov commented:
@timductive, thanks for your review.

> 1. There's no autocomplete on the Value field in the screenshot below. Obviously, I can guess that bytes should be numeric but if I try to type alpha characters nothing happens. It would be nice to either, have the kql dropdown with some proposed values or at least an error message when I attempt to input an invalid value.

Autocomplete is possible only for keywords.

https://user-images.githubusercontent.com/22456368/208373168-8a6aa4cc-4092-48b1-b50f-6b189fa631a0.mov

---

> 3. Sometimes the Preview shows the infinity symbol and sometimes its text.

That is a formatter for values of datetime format. If we'll change that formatter, it will affect all the Kibana.

---

> 5. There's some funkiness with pinning and switching between Discover and Lens. I can maybe recreate over zoom but if i just pin and unpin a bunch and switch between the apps, it seems like the state is getting confused..
Below in Lens i only have 1 pinned filter

I guess, that bug is related to the behavior of filters and can be reproducible on main. We have decided to fix it in a separate PR.

https://user-images.githubusercontent.com/22456368/208374105-5d769e1c-28cb-4552-b34d-52b9e5747dd2.mov

ppisljar reviewed (approved):
LGTM

michaelolo24 reviewed (approved):
Investigations codeowners review. Thanks! LGTM!
